nixos/modules/installer/tools/nixos-version.sh

```diff
@@ -7,25 +7,25 @@ case "$1" in
     exit 1
     ;;
   --hash|--revision)
-    if ! [[ @revision@ =~ ^[0-9a-f]+$ ]]; then
+    revision="$(jq --raw-output '.nixpkgsRevision // "null"' < /run/current-system/nixos-version.json)"
```

Comment:
I feel like this has been proposed before and was decided against. Can try and look for previous PR/issue if you want me to.

The reasoning was that this doesn't do the same thing. Previously, nixos-version executed from a nixos system closure always gives you the version of that system.

I understand and share the desire to reduce rebuilding these derivations, but this should be carefully be decided one,  because it 100% breaks existing behavior.

Reply:
Maybe this PR could just introduce the json file and a toggle for this script so that parties interested in getting rid of the constant rebuilds could remove the script, but still be able to get that information out by manually using `jq`.

Reply:
A JSON file is probably easier to consume than a shell script for automated tooling that wants the exact version anyway, right? I realize this is a compatibility-breaking change but it seems worth it in the long run.

Reply:
What is the use-case for executing a script from a non-active nixos closure? Do you have any visibility on where this is used? I can think of deployment tools maybe.
I do also think though that parsing a top-level JSON file that has this data seems easier and safer than relying on the output of a script that you need to go look for. It's not very obvious either that that script exists if you didn't already know about it.

Reply:
> a toggle for this script so that parties interested in getting rid of the constant rebuilds could remove the script

we should really turn this on by default. Rebuilding and restarting those handful of configs is creating noise and sometimes even breaking things.

Reply:
So a good way forward would be to guard this with a configuration switch for the time being while additionally adding a `json` file to the configuration for any tooling that wants to parse it. Then we can inform potential users of the old way and switch the default later?

Reply:
That sounds reasonable to me. If we have an agreement that this is the way forward, then I can implement it this way.

Reply:
Is there any way we can proactively surface this to tooling? Maybe print a warning to stderr when running `nixos-version` on an old version of the system?

Reply:
I guess that's a little circular in that it involves comparing versions :) But "store path of `$0` != store path of `/run/current-system`" might be sufficient.

Reply:
I pushed a new version addressing these concerns. The new behaviour is behind an option now (`system.installerTools.nixos-version.selfContained`), which defaults to `true`, giving the old behaviour. In that case we substitute the store path of the JSON file directly into the `nixos-version` script (hence why it's "self-contained").

When setting this to false, `nixos-version` looks at `/run/current-system/nixos-version.json` instead, and so won't be changing with every change to the nixpkgs or config revision hashes. This behaviour must be explicitly enabled for now.

I'm open to suggestions for the names, if you think something else would be more suitable. I also wonder whether we'd at some point want to switch this default, or not.

Reply:
@ajs124 @xaverdh @emilazy do you think we can move forward after the changes I made to incorporate your feedback?

Reply:
I think this is fine now! We can still discuss whether to change defaults in the future, but that should not delay this pr.

Reply:
Yeah I guess it would be best to wait for this to hit stable before introducing any kind of warning along the lines of what I suggested, since it could break tooling.